Added new image metric - UQI

ankitaS11 opened the pull request:
## What does this PR do?

Adds new Metric UQI as requested in the tracking issue #799

Notes: 

1. UQI will also be used in the future by `D_LAMBDA` and `D_S` metrics, both of which will help implement no reference metric. Please see my comment [here](https://github.com/PyTorchLightning/metrics/issues/799#issuecomment-1026592752) for a little more context.
2. There is similarity between the mathematical formula of SSIM and UQI which is also reflected in this PR (please see below for more context).

From the paper titled [Image quality assessment: from error visibility to structural similarity](https://ieeexplore.ieee.org/document/1284395):

> The “universal quality index” (UQI) defined in [6] and [7] corresponds to the special case that C1 = C2 = 0.

(where C1 and C2 are equivalent to k1 and k2 in the code base of SSIM)

I've also verified from the formulae of SSIM and UQI, that when k1 and k2 are 0 then the formulae are same. Hence, the additions are similar to what present in SSIM class/functions.

## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests? 

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Yes, I had a great time.

cc: @Borda @justusschock

ankitaS11 commented on torchmetrics/functional/image/uqi.py:
```diff
@@ -0,0 +1,246 @@
+        sigma: Standard deviation of the gaussian kernel
+        dtype: data type of the output tensor
+        device: device of the output tensor
+
+    Example:
+        >>> _gaussian(3, 1, torch.float, 'cpu')
+        tensor([[0.2741, 0.4519, 0.2741]])
+    """
+    dist = torch.arange(start=(1 - kernel_size) / 2, end=(1 + kernel_size) / 2, step=1, dtype=dtype, device=device)
+    gauss = torch.exp(-torch.pow(dist / sigma, 2) / 2)
+    return (gauss / gauss.sum()).unsqueeze(dim=0)  # (1, kernel_size)
+
+
+def _gaussian_kernel(
+    channel: int, kernel_size: Sequence[int], sigma: Sequence[float], dtype: torch.dtype, device: torch.device
+) -> Tensor:
+    """Computes 2D gaussian kernel.
+
+    Args:
+        channel: number of channels in the image
+        kernel_size: size of the gaussian kernel as a tuple (h, w)
+        sigma: Standard deviation of the gaussian kernel
+        dtype: data type of the output tensor
+        device: device of the output tensor
+
+    Example:
+        >>> _gaussian_kernel(1, (5,5), (1,1), torch.float, "cpu")
+        tensor([[[[0.0030, 0.0133, 0.0219, 0.0133, 0.0030],
+                  [0.0133, 0.0596, 0.0983, 0.0596, 0.0133],
+                  [0.0219, 0.0983, 0.1621, 0.0983, 0.0219],
+                  [0.0133, 0.0596, 0.0983, 0.0596, 0.0133],
+                  [0.0030, 0.0133, 0.0219, 0.0133, 0.0030]]]])
+    """
+
+    gaussian_kernel_x = _gaussian(kernel_size[0], sigma[0], dtype, device)
+    gaussian_kernel_y = _gaussian(kernel_size[1], sigma[1], dtype, device)
+    kernel = torch.matmul(gaussian_kernel_x.t(), gaussian_kernel_y)  # (kernel_size, 1) * (1, kernel_size)
+
+    return kernel.expand(channel, 1, kernel_size[0], kernel_size[1])
+
```
To the reviewers: These two functions for gaussian kernel were already present in the code of SSIM, and are copied from there. I think that we can directly import these functions from the SSIM file, but just wanted to confirm before doing this.

stancld replied:
Considering these two methods are used across multiple metrics, maybe we can move them into a separate helper file such as `torchmetrics/funcional/image/helper.py`? Any thoughts, @SkafteNicki @Borda?

justusschock replied:
yes, please reuse them :) Try to avoid code duplication whenever possible

justusschock replied:
yes, we should do so :)

ankitaS11 replied:
Thanks @stancld @justusschock, I have moved them in a new file `torchmetrics/functional/image/helper.py`. I hope this looks good now :)

ankitaS11 commented:
Hi @SkafteNicki @stancld 

This PR should be now ready for your first review. Can you please take a look?

ankitaS11 commented:
Looks like the [test](https://github.com/PyTorchLightning/metrics/runs/5074417691?check_suite_focus=true) doesn't work for PyTorch version less than 1.7. The issue seems to be with comparisons, I am a little confused because a similar test passes for [SSIM](https://github.com/PyTorchLightning/metrics/blob/master/tests/image/test_ssim.py#L137), but fails for UQI (only for old PyTorch versions). Can someone please help with this?

ashutoshml commented on torchmetrics/functional/image/uqi.py:
```diff
@@ -0,0 +1,245 @@
+    return _uqi_compute(preds, target, kernel_size, sigma, reduction, data_range)
+
+
+@deprecated(target=universal_image_quality_index, deprecated_in="0.7", remove_in="0.8", stream=_future_warning)
```
Since UQI is a new metric being introduced in v0.8, this deprecated function should be removed.

ashutoshml commented on torchmetrics/image/uqi.py:
```diff
@@ -0,0 +1,127 @@
+        return _uqi_compute(preds, target, self.kernel_size, self.sigma, self.reduction, self.data_range)
+
+
+class UQI(UniversalImageQualityIndex):
```
This class too is redundant, since UQI is a new metric

stancld commented on torchmetrics/__init__.py:
```diff
@@ -54,9 +54,11 @@
 from torchmetrics.image import (  # noqa: E402
     PSNR,
     SSIM,
+    UQI,
```
```suggestion
```

As @ashutoshml pointed out, `UQI` is a new metric for `v0.8` hence there is no need to introduce `UQI` next to `UniversalImageQualityIndex`

justusschock replied:
```suggestion
```

stancld commented on torchmetrics/__init__.py:
```diff
@@ -176,6 +178,8 @@
     "SumMetric",
     "SymmetricMeanAbsolutePercentageError",
     "TranslationEditRate",
+    "UQI",
+    "UniversalImageQualityIndex",
```
```suggestion
    "UniversalImageQualityIndex",
```

stancld commented on torchmetrics/functional/__init__.py:
```diff
@@ -41,6 +41,7 @@
     ssim,
     structural_similarity_index_measure,
 )
+from torchmetrics.functional.image.uqi import universal_image_quality_index, uqi
```
```suggestion
from torchmetrics.functional.image.uqi import universal_image_quality_index
```

justusschock replied:
```suggestion
from torchmetrics.functional.image.uqi import universal_image_quality_index

stancld commented on torchmetrics/functional/__init__.py:
```diff
@@ -149,6 +150,8 @@
     "stat_scores",
     "symmetric_mean_absolute_percentage_error",
     "translation_edit_rate",
+    "uqi",
+    "universal_image_quality_index",
```
```suggestion
    "universal_image_quality_index",
```

stancld commented on torchmetrics/functional/image/__init__.py:
```diff
@@ -18,3 +18,4 @@
     ssim,
     structural_similarity_index_measure,
 )
+from torchmetrics.functional.image.uqi import universal_image_quality_index, uqi  # noqa: F401
```
```suggestion
from torchmetrics.functional.image.uqi import universal_image_quality_index  # noqa: F401
```

justusschock replied:
```suggestion
from torchmetrics.functional.image.uqi import universal_image_quality_index  # noqa: F401
```

stancld commented on torchmetrics/functional/image/uqi.py:
```diff
@@ -0,0 +1,245 @@
+from typing import Optional, Sequence, Tuple, Union
+
+import torch
+from deprecate import deprecated, void
```
```suggestion
```

stancld commented on torchmetrics/functional/image/uqi.py:
```diff
@@ -0,0 +1,245 @@
+    target: Tensor,
+    kernel_size: Sequence[int] = (11, 11),
+    sigma: Sequence[float] = (1.5, 1.5),
+    reduction: str = "elementwise_mean",
```
```suggestion
    reduction: Literal["elementwise_mean", "sum", "none"] = "elementwise_mean",
```

and use `from typing_extensions import Literal`

stancld commented on torchmetrics/functional/image/uqi.py:
```diff
@@ -0,0 +1,245 @@
+    target: Tensor,
+    kernel_size: Sequence[int] = (11, 11),
+    sigma: Sequence[float] = (1.5, 1.5),
+    reduction: str = "elementwise_mean",
```
^^
```suggestion
    reduction: Literal["elementwise_mean", "sum", "none"] = "elementwise_mean",
```

stancld commented on torchmetrics/functional/image/uqi.py:
```diff
@@ -0,0 +1,245 @@
+        tensor(0.9216)
+
+    References:
+    [1] Zhou Wang and A. C. Bovik, "A universal image quality index," in IEEE Signal Processing Letters, vol. 9,
+    no. 3, pp. 81-84, March 2002, doi: 10.1109/97.995823.
```
Please add this reference into `docs/source/links.rst`

ankitaS11 replied:
Thank you @stancld, I have fixed this :)

stancld commented on torchmetrics/image/__init__.py:
```diff
@@ -19,6 +19,7 @@
     MultiScaleStructuralSimilarityIndexMeasure,
     StructuralSimilarityIndexMeasure,
 )
+from torchmetrics.image.uqi import UQI, UniversalImageQualityIndex  # noqa: F401
```
```suggestion
from torchmetrics.image.uqi import UniversalImageQualityIndex  # noqa: F401
```

stancld commented on torchmetrics/image/uqi.py:
```diff
@@ -0,0 +1,127 @@
+from typing import Any, List, Optional, Sequence
+
+import torch
+from deprecate import deprecated, void
```
```suggestion
```

justusschock commented on torchmetrics/functional/image/uqi.py:
```diff
@@ -0,0 +1,245 @@
+            f" Got preds: {preds.dtype} and target: {target.dtype}."
+        )
+    _check_same_shape(preds, target)
+    if len(preds.shape) != 4:
```
Should we wait with this for #818 and also prepare this PR for 3D directly? :)

ankitaS11 replied:
I took a quick look at the PR for 3D ssim, but I'm not sure how long it might take for that PR to land. Will it be okay if I create a follow-up PR to make necessary changes after 3D SSIM once it's merged? (I'm having a few quizzes and exams in the coming week, so would love to complete this PR sooner, so that this doesn't go stale and doesn't block anyone else). :) Also, once this is merged, it will also help me implement `D_S` and `D_LAMBDA` and then QNR metric.

justusschock commented on torchmetrics/functional/image/uqi.py:
```diff
@@ -0,0 +1,245 @@
+    return _uqi_compute(preds, target, kernel_size, sigma, reduction, data_range)
+
+
+@deprecated(target=universal_image_quality_index, deprecated_in="0.7", remove_in="0.8", stream=_future_warning)
+def uqi(
+    preds: Tensor,
+    target: Tensor,
+    kernel_size: Sequence[int] = (11, 11),
+    sigma: Sequence[float] = (1.5, 1.5),
+    reduction: str = "elementwise_mean",
+    data_range: Optional[float] = None,
+) -> Tensor:
+    """Computes Universal Image Quality Index.
+
+    Example:
+        >>> preds = torch.rand([16, 1, 16, 16])
+        >>> target = preds * 0.75
+        >>> uqi(preds, target)
+        tensor(0.9216)
+    """
+    return void(preds, target, kernel_size, sigma, reduction, data_range)
```
```suggestion
```

as @ashutoshml suggested :)

justusschock commented on torchmetrics/image/uqi.py:
```diff
@@ -0,0 +1,127 @@
+        return _uqi_compute(preds, target, self.kernel_size, self.sigma, self.reduction, self.data_range)
+
+
+class UQI(UniversalImageQualityIndex):
+    """Computes Universal Image Quality Index (UQI_).
+
+    .. deprecated:: v0.7
+        Use :class:`torchmetrics.UniversalImageQualityIndex`. Will be removed in v0.8.
+
+    Example:
+        >>> preds = torch.rand([16, 1, 16, 16])
+        >>> target = preds * 0.75
+        >>> uqi = UQI()
+        >>> uqi(preds, target)
+        tensor(0.9216)
+    """
+
+    @deprecated(target=UniversalImageQualityIndex, deprecated_in="0.7", remove_in="0.8", stream=_future_warning)
+    def __init__(
+        self,
+        kernel_size: Sequence[int] = (11, 11),
+        sigma: Sequence[float] = (1.5, 1.5),
+        reduction: str = "elementwise_mean",
+        data_range: Optional[float] = None,
+        compute_on_step: bool = True,
+        dist_sync_on_step: bool = False,
+        process_group: Optional[Any] = None,
+    ) -> None:
+        void(kernel_size, sigma, reduction, data_range, compute_on_step, dist_sync_on_step, process_group)
```
```suggestion
```

as suggested by @ashutoshml

justusschock commented on torchmetrics/__init__.py:
```diff
@@ -176,6 +178,8 @@
     "SumMetric",
     "SymmetricMeanAbsolutePercentageError",
     "TranslationEditRate",
+    "UQI",
```
```suggestion
```

stancld reviewed (commented):
Hi @ankitaS11, thank you a lot for this nice contribution :]

I left a few comments, but it looks really promising. Please. also add references to this new metric into `docs/source/references/functional.rst` and `docs/source/references/modules.rst` so that they will be included in docs :] Also, you can already update `CHANGELOG.md` that this new metrics has been added.

I'll have a look at the tests later today.

justusschock commented on torchmetrics/functional/__init__.py:
```diff
@@ -149,6 +150,8 @@
     "stat_scores",
     "symmetric_mean_absolute_percentage_error",
     "translation_edit_rate",
+    "uqi",
```
```suggestion
```

justusschock commented on torchmetrics/functional/image/uqi.py:
```diff
@@ -0,0 +1,224 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Optional, Sequence, Tuple, Union
```
```suggestion
from typing import Optional, Sequence, Tuple, Union, Literal
```

justusschock commented on torchmetrics/image/uqi.py:
```diff
@@ -0,0 +1,98 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, List, Optional, Sequence
```
```suggestion
from typing import Any, List, Optional, Sequence, Literal
```

ashutoshml replied:
`Literal` might need import from `typing_extensions`

justusschock commented on torchmetrics/image/uqi.py:
```diff
@@ -0,0 +1,98 @@
+        self,
+        kernel_size: Sequence[int] = (11, 11),
+        sigma: Sequence[float] = (1.5, 1.5),
+        reduction: str = "elementwise_mean",
```
```suggestion
        reduction: Literal["elementwise_mean", "sum", "none"] = "elementwise_mean",
```

see comments from @stancld

justusschock commented:
> Looks like the [test](https://github.com/PyTorchLightning/metrics/runs/5074417691?check_suite_focus=true) doesn't work for PyTorch version less than 1.7. The issue seems to be with comparisons, I am a little confused because a similar test passes for [SSIM](https://github.com/PyTorchLightning/metrics/blob/master/tests/image/test_ssim.py#L137), but fails for UQI (only for old PyTorch versions). Can someone please help with this?

@ankitaS11 for tensors in general, you should use `torch.allclose` to compare them :) Can you give this a try and see if it helps?

ashutoshml commented on torchmetrics/image/uqi.py:
```diff
@@ -0,0 +1,98 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, List, Literal, Optional, Sequence
```
`Literal` should be imported from `typing_extensions`

justusschock replied:
```suggestion
from typing import Any, List, Optional, Sequence
from typing_extensions import Literal
```

justusschock commented on torchmetrics/functional/image/uqi.py:
```diff
@@ -0,0 +1,224 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Literal, Optional, Sequence, Tuple, Union
```
```suggestion
from typing import Optional, Sequence, Tuple, Union
from typing_extensions import Literal
```

ashutoshml commented on torchmetrics/image/uqi.py:
```diff
@@ -0,0 +1,99 @@
+
+from torchmetrics.functional.image.uqi import _uqi_compute, _uqi_update
+from torchmetrics.metric import Metric
+from torchmetrics.utilities import _future_warning, rank_zero_warn
```
`_future_warning` is redundant.

ankitaS11 replied:
Thanks @ashutoshml, this is fixed now.

ashutoshml commented on torchmetrics/functional/image/uqi.py:
```diff
@@ -0,0 +1,225 @@
+from torch.nn import functional as F
+from typing_extensions import Literal
+
+from torchmetrics.utilities import _future_warning
```
`_future_warning` is redundant.

ankitaS11 commented:
Hi, @justusschock @stancld @ashutoshml - Thank you so much for reviewing this PR. This was my first one in `metrics`, and I learned a lot while working on it. I've addressed the review comments, and hopefully the tests should pass now (in case they don't, I'll check them as soon as possible). Can you please take a look again whenever possible? :)

ashutoshml commented on docs/source/references/functional.rst:
```diff
@@ -270,6 +270,12 @@ peak_signal_noise_ratio [func]
     :noindex:
 
+universal_image_quality_index [func]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
The length of `~` here should be equal to the length of the preceeding string 
`universal_image_quality_index [func]`

ankitaS11 replied:
Oh, my bad. Thanks for pointing this out, I have fixed this now :)

ashutoshml commented on docs/source/references/modules.rst:
```diff
@@ -404,6 +404,12 @@ StructuralSimilarityIndexMeasure
     :noindex:
 
+UniversalImageQualityIndex
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Similar length argument for this.

stancld commented on requirements.txt:
```diff
@@ -2,3 +2,4 @@ numpy>=1.17.2
 torch>=1.3.1
 pyDeprecate==0.3.*
 packaging
+typing_extensions
```
```suggestion
```

`typing_extensions` is among the python built-in packages since `python>=3.5` thus there's no need to specify this package in `requirements.txt`.

stancld commented on tests/image/test_uqi.py:
```diff
@@ -0,0 +1,174 @@
+
+from tests.helpers import seed_all
+from tests.helpers.testers import BATCH_SIZE, NUM_BATCHES, MetricTester
+from torchmetrics.functional import universal_image_quality_index
+from torchmetrics.image import UniversalImageQualityIndex
```
```suggestion
from torchmetrics.functional.image.uqi import universal_image_quality_index
from torchmetrics.image.uqi import UniversalImageQualityIndex
```
We usually use full importing path in test files

stancld reviewed (approved):
Just two minor comments, otherwise LGTM

ashutoshml commented on torchmetrics/image/uqi.py:
```diff
@@ -0,0 +1,99 @@
+
+
+class UniversalImageQualityIndex(Metric):
+    """Computes Universal Image Quality Index (UQI_).
```
`UQI_` -> `UniversalImageQualityIndex_` (Since this is the name in `links.rst`)

ankitaS11 replied:
Thank you @ashutoshml, this is fixed now :)

ashutoshml commented on torchmetrics/image/uqi.py:
```diff
@@ -0,0 +1,99 @@
+            process_group=process_group,
+        )
+        rank_zero_warn(
+            "Metric `UQI` will save all targets and"
```
`UQI` -> `UniversalImageQualityIndex`

ashutoshml commented on torchmetrics/functional/image/uqi.py:
```diff
@@ -0,0 +1,179 @@
+        data_range: Range of the image. If ``None``, it is determined from the image (max - min)
+
+    Return:
+        Tensor with UQI score
```
UQI -> UniversalImageQualityIndex

ashutoshml commented on torchmetrics/image/uqi.py:
```diff
@@ -0,0 +1,99 @@
+        data_range: Range of the image. If ``None``, it is determined from the image (max - min)
+
+    Return:
+        Tensor with UQI score
```
`UQI` -> `UniversalImageQualityIndex`

Borda commented on torchmetrics/image/uqi.py:
```diff
@@ -0,0 +1,99 @@
+
+    preds: List[Tensor]
+    target: List[Tensor]
+    higher_is_better = True
```
```suggestion
    higher_is_better: bool = True
```
